Userland/Libraries/LibJS/Bytecode/Pass/AllocateRegisters.cpp

```diff
@@ -0,0 +1,173 @@
+        auto& inst = *i;
+
+        for (auto& reg : inst.write_registers()) {
+            m_is_live[reg->index()] = true;
+            m_live_ranges[reg->index()].reg = reg->index();
+            m_live_ranges[reg->index()].start = { block_id(basic_block), i.offset() };
+            m_live_ranges[reg->index()].finish = { 0, 0 };
+        }
+
+        for (auto& reg : inst.read_registers()) {
+            if (max_bb < m_live_ranges[reg->index()].finish.basic_block)
+                continue;
+
+            if (block_id(basic_block) == max_bb) {
+                m_live_ranges[reg->index()].finish = { block_id(basic_block), i.offset() };
+            } else {
+                // The maximum size_t is equivalent to the end of the basic block.
+                m_live_ranges[reg->index()].finish = { max_bb, NumericLimits<size_t>::max() };
+            }
+        }
+    }
+}
+
+void AllocateRegisters::find_live_ranges(const BasicBlock* basic_block, size_t max_bb)
+{
+    if (m_live_range_path.contains_slow(basic_block))
+        return;
+
+    m_live_range_path.append(basic_block);
+    max_bb = max(block_id(basic_block), max_bb);
+
+    find_block_range(basic_block, max_bb);
+
+    const auto& children = m_executable->cfg.value().get(basic_block);
+    if (children.has_value()) {
+        for (auto& child : children.value()) {
+            find_live_ranges(child, max_bb);
+        }
+    }
+
+    m_live_range_path.take_last();
+}
+
+Vector<Register> AllocateRegisters::rename_registers()
+{
+    Vector<Register> rename;
+    for (size_t i = 0; i < m_executable->executable.number_of_registers; ++i) {
+        rename.append(Register(i));
+    }
+
+    Vector<size_t> active;
+    active.resize(m_executable->executable.number_of_registers);
+    for (auto& i : active) {
+        i = 0;
+    }
+
+    quick_sort(m_live_ranges, [](LiveRange a, LiveRange b) { return a.start < b.start; });
+
+    for (size_t i = 2; i < m_live_ranges.size(); ++i) {
+        auto& timestamp = m_live_ranges[i].start;
+
+        // Expire old intervals.
+        for (size_t j = 2; j < i; ++j) {
+            auto reg = rename[m_live_ranges[j].reg].index();
+            if (m_live_ranges[j].finish < timestamp && active[reg] == j) {
+                active[reg] = 0;
+            }
+        }
+
+        // Select a register.
+        for (size_t j = 2; j <= i; ++j) {
+            if (!active[j]) {
+                rename[m_live_ranges[i].reg] = Register(j);
+                active[j] = i;
+                break;
+            }
+        }
+    }
+
+    return rename;
+}
+
+void AllocateRegisters::apply_register_rename(Vector<Register> const& rename)
+{
+    for (auto& basic_block : m_executable->executable.basic_blocks) {
+        for (auto i = InstructionStreamIterator(basic_block.instruction_stream()); !i.at_end(); ++i) {
+            auto& inst = *i;
+            for (auto reg : inst.write_registers()) {
+                *reg = rename[reg->index()];
```

Comment:
It'd be kinda nicer if we could ask the instruction to replace the old register with the new one, tbh
Could mirror `replace_references(BasicBlock....)`.